gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java

```diff
@@ -162,6 +175,90 @@ public void testBackendConfiguration()
+
+        Request initiateRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + oauthInitiatePath)
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "custom")
+                        .build();
+        Response initiateResponse = httpClient.newCall(initiateRequest).execute();
+        assertThat(isNullOrEmpty(initiateResponse.header("set-cookie"))).isFalse();
+
+        Request callbackRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + oauthCallbackPath)
+                        .post(requestBody)
+                        .addHeader("Cookie", initiateResponse.header("set-cookie"))
+                        .build();
+        Response callbackResponse = httpClient.newCall(callbackRequest).execute();
+        assertThat(callbackResponse.body().string()).isEqualTo(oauthCallbackResponse);
+
+        Request logoutRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + "/custom/logout")
+                        .post(requestBody)
+                        .addHeader("Cookie", initiateResponse.header("set-cookie"))
+                        .build();
+        Response logoutResponse = httpClient.newCall(logoutRequest).execute();
+
+        assertThat(isNullOrEmpty(logoutResponse.header("set-cookie"))).isFalse();
+    }
+
+    @Test
+    public void testCookieSigning()
+            throws IOException
+    {
+        OkHttpClient httpClient = new OkHttpClient.Builder()
+                .connectTimeout(10, TimeUnit.SECONDS)
+                .writeTimeout(10, TimeUnit.SECONDS)
+                .readTimeout(60, TimeUnit.SECONDS)
+                .build();
+        String oauthInitiateBody = "anything";
+        RequestBody requestBody =
+                RequestBody.create(
+                        MediaType.parse("application/json; charset=utf-8"), oauthInitiateBody);
+
+        Request initiateRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + oauthInitiatePath)
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "custom")
+                        .build();
+        Response initiateResponse = httpClient.newCall(initiateRequest).execute();
+        String cookieHeader = initiateResponse.header("set-cookie");
+        assertThat(isNullOrEmpty(cookieHeader)).isFalse();
+        GatewayCookie cookie = GatewayCookie.fromCookie(new Cookie(cookieHeader.split("=")[0], cookieHeader.split("=")[1].split(";")[0]));
+        assertThat(isNullOrEmpty(cookie.getSignature())).isFalse();
+
+        // Tamper with values. This will cause the cookie to be ignored because its values will not match the signature,
+        // causing the request will be routed to the adhoc backend
+        cookie.setTs(System.currentTimeMillis());
```

Comment:
This test sometimes fail in my local environment.
Shell we set another field other than ts?

```
[ERROR] Failures: 
[ERROR]   TestGatewayHaMultipleBackend.testCookieSigning:259 
expected: 500
 but was: 200
[INFO] 
[ERROR] Tests run: 95, Failures: 1, Errors: 0, Skipped: 0
```

Reply:
`setTs` is the only setter, since all the other fields are final. `gatewayCookie.setTs(gatewayCookie.getTs() + 1000)` should fail deterministically - do you think this is ok or should we construct a new cookie with a different backend?

Reply:
`gatewayCookie.setTs(gatewayCookie.getTs() + 1000)` should be good.